Library/Homebrew/os/mac/cache_store.rb

```diff
@@ -0,0 +1,137 @@
+require "dbm"
+require "json"
+
+#
+# `DatabaseCache` acts as an interface to a persistent storage mechanism
+# residing in the `HOMEBREW_CACHE`
+#
+class DatabaseCache
+  # Users have read and write, but not execute permissions
+  DATABASE_MODE = 0666
+
+  # Opens and yields a database in read/write mode
+  #
+  # DBM::WRCREAT: Creates the database if it does not already exist
+  def initialize(name)
+    @db = DBM.open("#{HOMEBREW_CACHE}/#{name}.db", DATABASE_MODE, DBM::WRCREAT)
+    yield(@db)
+    @db.close
+  end
+end
+
+#
+# `CacheStore` provides methods to mutate and fetch data from a persistent
+# storage mechanism
+#
+class CacheStore
+  def initialize(database_cache)
+    @database_cache = database_cache
+  end
+
+  # Inserts new values or updates existing cached values to persistent storage
+  # mechanism
+  def update!(*)
+    raise NotImplementedError
+  end
+
+  # Fetches cached values in persistent storage according to the type of data
+  # stored
+  def fetch_type(*)
+    raise NotImplementedError
+  end
+
+  # Deletes data from the cache based on a condition defined in a concrete class
+  def flush_cache!
+    raise NotImplementedError
+  end
+
+  protected
+
+  attr_reader :database_cache
+
+  # Parses `DBM` stored `String` into ruby `Hash`
+  #
+  # DBM stores ruby objects as a ruby `String`. Hence, when fetching the data,
+  # to convert the ruby string back into a ruby `Hash`, the string is converted
+  # into a JSON compatible string, where it may be parsed by the JSON.parse
+  # function
+  def string_to_hash(string)
+    JSON.parse(string.gsub("=>", ":"))
+  end
+end
+
+#
+# `LinkageStore` provides methods to fetch and mutate linkage-specific data used
+# by the `brew linkage` command
+#
+class LinkageStore < CacheStore
+  HASH_LINKAGE_TYPES = [:brewed_dylibs, :reverse_links].freeze
+
+  def initialize(keg_name, database_cache)
+    @keg_name = keg_name
+    super(database_cache)
+  end
+
+  def update!(
+    path_values: {
+      system_dylibs: %w[],
+      variable_dylibs: %w[],
+      broken_dylibs: %w[],
+      indirect_deps: %w[],
+      undeclared_deps:  %w[],
+      unnecessary_deps: %w[],
+    },
+    hash_values: {
+      brewed_dylibs: {},
+      reverse_links: {},
+    }
+  )
+    database_cache[keg_name] = {
+      "path_values" => format_path_values(path_values),
+      "hash_values" => format_hash_values(hash_values),
+    }
+  end
+
+  def fetch_type(type)
+    if HASH_LINKAGE_TYPES.include?(type)
+      fetch_hash_values(type: type)
+    else
+      fetch_path_values(type: type)
+    end
+  end
+
+  def flush_cache!
+    database_cache.delete(keg_name)
+  end
+
+  private
+
+  attr_reader :keg_name
+
+  def fetch_path_values(type:)
+    return [] if !database_cache.key?(keg_name) || database_cache[keg_name].nil?
```

Comment:
When would `.key?` be true but `[]` be false?

Reply:
I seem to recall coming across a situation in which the key existed, but the value was `nil` - causing a runtime error. I can remove the right-hand side of the boolean expression, and test to see if there are any issues.

Reply:
I did some investigation, and I remember why I added the extra condition. There's an edge case where you could get:

```
Error: undefined method `[]' for nil:NilClass
Please report this bug:
  https://docs.brew.sh/Troubleshooting.html
/usr/local/Homebrew/Library/Homebrew/os/mac/cache_store.rb:117:in `fetch_array_values'
```

Reply:
Nevermind, that issue was unrelated. Apologies for jumping to a premature conclusion. I think it's fine to remove the other condition.